src/main/java/org/templegalaxia/patterns/acjs/Spoken.java

```diff
@@ -0,0 +1,216 @@
+        e.printStackTrace();
+      } finally {
+        // Backup / default
+        names.add("Larry Harvey");
+      }
+    }
+
+    Collections.shuffle(names);
+  }
+
+  public void newPulser(String message) {
+    double dotLength = dot.getValue();
+    dotLength += (Math.random() - 0.5) * dotVariance.getValue();
+
+    Collections.shuffle(availableLights);
+    int light = availableLights.remove(0);
+
+    MorsePulser pulser = new MorsePulser(lx, light, message, dotLength);
+    pulsers.add(pulser);
+    addLayer(pulser);
+  }
+
+  public void run(double deltaMs) {
+    // Give each unlit 'slot' a chance to be born.
+    for (int i = 0; i < (int) density.getValue() - pulsers.size(); i++) {
+      if (names.size() == 0) loadNames(); // Reload names if we're out. XXX: Performance?
+
+      if (rand.nextDouble() < birthRate.getValue() * deltaMs / 1000) {
+        newPulser(names.remove(0));
+      }
+    }
+
+    // Garbage collect completed pulsers
+    // TODO: Do this with a callback on finish or something.
+    for (Iterator<MorsePulser> iterator = pulsers.iterator(); iterator.hasNext(); ) {
+      MorsePulser pulse = iterator.next();
+      if (pulse.isFinished()) {
+        availableLights.add(pulse.getLightIndex());
+        iterator.remove();
+      }
+    }
+  }
+
+  public class MorsePulser extends LXLayer {
+    private final int lightIndex;
+    private final String message;
+    private final String morsePattern;
+
+    private final LXPeriodicModulator dotModulator;
+    private final LXPeriodicModulator dashModulator;
+
+    private int currentCharacter = 0;
+    private boolean finished = false;
+
+    public MorsePulser(LX lx, int lightIndex, String message, double dotLength) {
+      super(lx);
+
+      message = message.toUpperCase();
+
+      StringBuffer morse = new StringBuffer();
+      for (int i = 0; i < message.length(); i++) {
+        if (morseLookup.containsKey(message.charAt(i))) {
+          if (i != 0) {
+            morse.append(" ");
+          }
+
+          morse.append(morseLookup.get(message.charAt(i)));
+        }
+      }
+
+      this.lightIndex = lightIndex;
+      this.message = message;
+      this.morsePattern = morse.toString();
+
+      dotModulator = new SinLFO(0, 100, dotLength).setLooping(false);
+      dashModulator = new SinLFO(0, 100, dotLength * DASH_MULTIPLIER).setLooping(false);
+
+      addModulator(dotModulator);
+      addModulator(dashModulator);
+
+      currentModulator().start();
+    }
+
+    LXPeriodicModulator currentModulator() {
+      if (morsePattern.charAt(currentCharacter) == '-') return dashModulator;
+      else return dotModulator;
+    }
+
+    @Override
+    public void run(double v) {
+      if (finished) return;
+
+      if (currentModulator().finished()) {
+        currentCharacter++;
+
+        if (currentCharacter >= morsePattern.length()) {
+          finished = true;
+          return;
+        }
+
+        currentModulator().start();
```

Comment:
I think this is fine, but currentModulator().trigger() may be more semantically appropriate (start means resume running from wherever it was, trigger() means start a new cycle from the beginning).

Reply:
Great to know. I feel like I tried trigger first and it wasn't working, but that was likely while something else was broken.